server/src/main/java/io/spine/server/commandbus/CommandErrorHandler.java

```diff
@@ -70,22 +71,156 @@ public static CommandErrorHandler with(RejectionBus rejectionBus) {
      * a {@linkplain io.spine.base.ThrowableMessage command rejection}, the {@link Rejection} is
      * {@linkplain RejectionBus#post(Rejection) posted} to the {@code RejectionBus}. Otherwise,
      * the given {@code exception} is thrown.
+     *
+     * @return the result of the error handing
      */
-    public void handleError(CommandEnvelope envelope, RuntimeException exception) {
+    @CanIgnoreReturnValue
+    public HandledError handleError(CommandEnvelope envelope, RuntimeException exception) {
         checkNotNull(envelope);
         checkNotNull(exception);
         if (causedByRejection(exception)) {
-            Rejection rejection = rejectWithCause(envelope.getCommand(), exception);
-            rejectionBus.post(rejection);
+            return handleRejection(envelope, exception);
+        } else {
+            return handleRuntime(envelope, exception);
+        }
+    }
+
+    private HandledError handleRejection(CommandEnvelope envelope, RuntimeException exception) {
+        Rejection rejection = rejectWithCause(envelope.getCommand(), exception);
+        rejectionBus.post(rejection);
+        markRejected(envelope, rejection);
+        return HandledError.forRejection();
+    }
+
+    private void markRejected(CommandEnvelope command, Rejection rejection) {
+        CommandId commandId = command.getId();
+        MarkCommandAsRejected systemCommand = MarkCommandAsRejected
+                .newBuilder()
+                .setId(commandId)
+                .setRejection(rejection)
+                .build();
+        postSystem(systemCommand);
+    }
+
+    private HandledError handleRuntime(CommandEnvelope envelope, RuntimeException exception) {
+        if (isPreProcessed(exception)) {
+            return HandledError.forPreProcessed();
         } else {
             log().error(format("Error dispatching command (class: %s id: %s).",
                                envelope.getMessage().getClass(),
                                Stringifiers.toString(envelope.getId())),
                         exception);
-            throw exception;
+            Error error = Errors.causeOf(exception);
+            markErrored(envelope, error);
+            return HandledError.forRuntime(exception);
         }
     }
 
+    private void markErrored(CommandEnvelope command, Error error) {
+        CommandId commandId = command.getId();
+        MarkCommandAsErrored systemCommand = MarkCommandAsErrored
+                .newBuilder()
+                .setId(commandId)
+                .setError(error)
+                .build();
+        postSystem(systemCommand);
+    }
+
+    private void postSystem(Message systemCommand) {
+        systemGateway.postCommand(systemCommand);
+    }
+
+    private static boolean isPreProcessed(RuntimeException exception) {
+        return exception instanceof CommandDispatchingException;
+    }
+
+    /**
+     * A result of an error handling.
+     */
+    public static final class HandledError {
```

Comment:
Please speak more of this entity.

Its current documentation says it's a value object. In fact it serves other reasons as well.